compiler/src/iree/compiler/Dialect/Stream/Transforms/UnifyEncodingForGlobals.cpp

```diff
@@ -270,9 +271,102 @@ GlobalEncodingAnalyzer::traceToSourceGlobal(Value value) {
+       llvm::enumerate(dispatchOp.getMixedOperands(),
+                       dispatchOp.getOperandEncodings().getValue())) {
+    Type type = cast<TypeAttr>(typeAttr).getValue();
+    if (!isa<IREE::Stream::AffinityTypeInterface>(type)) {
+      newOperandEncodings.push_back(typeAttr);
+      continue;
+    }
+    if (!operandUpdates.contains(idx)) {
+      newOperandEncodings.push_back(typeAttr);
+      continue;
+    }
+    Attribute newEncoding = operandUpdates.lookup(idx);
+    auto tensorType = cast<RankedTensorType>(type);
+    newOperandEncodings.push_back(
+        TypeAttr::get(tensorType.cloneWithEncoding(newEncoding)));
+    LDBG() << "  Updated dispatch operand encoding at index " << idx << " to "
+           << newEncoding;
+  }
+  dispatchOp.setOperandEncodingsAttr(
+      ArrayAttr::get(dispatchOp.getContext(), newOperandEncodings));
+}
+
+// Applies all cached encoding updates to tensor ops.
+static void applyTensorEncodingUpdates(TensorEncodingUpdates &updates) {
+  for (auto &[op, operandUpdates] : updates) {
+    // TODO: Handle other TensorPhaseOp ops (TensorFillOp, etc.) via TypeSwitch.
+    if (auto dispatchOp = dyn_cast<TensorDispatchOp>(op)) {
+      updateTensorDispatchOp(dispatchOp, operandUpdates);
+    }
+  }
+}
+
+// Collects updates for stream tensor ops by walking from global loads.
+static void collectUpdatesForStreamTensorOps(Explorer &explorer,
+                                             EncodedGlobalInfo &encodedInfo,
+                                             Attribute newEncoding,
+                                             TensorEncodingUpdates &updates) {
+  StringRef globalName = encodedInfo.encodedGlobal.getGlobalName();
+  LDBG() << "  Collecting updates for global: " << globalName;
+
+  // Get loads from the explorer's global info.
+  const Explorer::GlobalInfo *globalInfo =
+      explorer.queryGlobalInfoFrom(globalName, encodedInfo.encodedGlobal);
+  assert(globalInfo &&
+         "expected global info to be present in explorer for encoded global");
+  SmallVector<Value> worklist;
+  for (auto loadOp : globalInfo->getLoads()) {
+    worklist.push_back(loadOp.getLoadedGlobalValue());
+  }
+
+  while (!worklist.empty()) {
+    Value value = worklist.pop_back_val();
+    explorer.walkTransitiveUses(value, [&](OpOperand &operand) {
+      Operation *user = operand.getOwner();
+      if (auto cloneOp = dyn_cast<AsyncCloneOp>(user)) {
+        LDBG() << "      Following clone: " << cloneOp;
+        worklist.push_back(cloneOp.getResult());
+        return WalkResult::advance();
+      }
+
+      // Only stream tensor ops need to be updated. Skip other operations.
+      if (!user->hasTrait<OpTrait::IREE::Stream::TensorPhaseOp>()) {
+        return WalkResult::advance();
+      }
+
+      // TODO: Handle other tensor phase ops (TensorFillOp, etc.)
+      auto dispatchOp = dyn_cast<IREE::Stream::TensorDispatchOp>(user);
+      if (!dispatchOp) {
+        return WalkResult::advance();
```

Comment:
Is it correct to skip other `TensorPhaseOp` here? Wouldn't it be more correct to bail here if encounter a `TensorPhaseOp` that isn't handled yet?